### PR TITLE
test positivity provenance in API

### PR DESCRIPTION
**Pull request by TomGoBravo**
This PR follows up on https://github.com/covid-projections/covid-data-model/pull/728 to add provenance information.

* Populates provenance information for test positivity in the API output

## Tested

Ran a snapshot and looked at output such as https://data.covidactnow.org/snapshot/1283/v2/county/53033.timeseries.json where I see
```
metrics: {testPositivityRatio: null, testPositivityRatioDetails: {source: "other"},…}
```

### CODE REVIEW

File: libs/datasets/timeseries.py
Comment (mikelehen) [inline libs/datasets/timeseries.py:112]: `latest` and `provenance` are going to have columns outside of the `columns` subset specified, right?  But maybe we don't care?  Perhaps at least add a comment.

File: libs/datasets/timeseries.py
Comment (mikelehen) [inline libs/datasets/timeseries.py:486]: I was curious in what case we would be loading a CSV but have already-parsed provenance to go along with it...  It turns out we don't really.  But in tests, we load the provenance from a (string) CSV and then pass the _parsed_ version (as `pd.Series`) into from_csv() along with the _unparsed_ primary CSV (as just a `TextIO`).

This seems like a bit arbitrary / inconsistent.  Why not have `provenance` be `Optional[TextIO]` and always do the CSV parsing inside this method?  Or for full symmetry between the primary CSV and the provenance CSV, you could do `Optional[Union[pathlib.Path, TextIO]]` but that's overly flexible since we never need to pass in a Path right now.  🤷

File: api/can_api_definition.py
Comment (mikelehen) [inline api/can_api_definition.py:84]: Please make sure we don't ship this in the API as-is. 😁

**Review by mikelehen — COMMENTED** [review]
Left a few comments, but this seems reasonable to me.  I can review in more detail once this is ready to go (and merged with latest master so it doesn't include the diffs from the test positivity PR).

A couple high-level thoughts:
1. Provence is essentially just a per-location/column piece of metadata, right?  I wonder if we should be thinking about creating that abstraction instead (column metadata) instead of baking this specific use of column metadata (provenance) into our data structures and naming.
2. On that note, test_positivity already has kinda' 2 forms of provenance, right? a) what data source did the data come from ("X dashboard via covid-tracking"), and b) which calculation did we use to calculate the final metric value ("positive_tests_viral / total_tests_viral").  Do you have a sense of how we would represent that?  If we have a single provenance field per location/column, do we have to come up with some sort of string-encoding of all of that (a JSON blob or something)?

Also, just FYI- When I run this branch locally right now I get tons of "Unable to find TestPositivityRatioMethod location_id=iso1:us#iso2:us-sd#fips:46003" errors which I don't get in master.  But maybe that'll be fixed by a merge.

File: libs/datasets/timeseries.py
Comment (TomGoBravo) [reply libs/datasets/timeseries.py]: Yes, that sucked. Fixed with https://github.com/covid-projections/covid-data-model/pull/757

File: libs/datasets/timeseries.py
Comment (TomGoBravo) [reply libs/datasets/timeseries.py]: Maybe, but probably not soon.

**Comment by TomGoBravo** [comment]
> Left a few comments, but this seems reasonable to me. I can review in more detail once this is ready to go (and merged with latest master so it doesn't include the diffs from the test positivity PR).
> 
> A couple high-level thoughts:
> 
> 1. Provence is essentially just a per-location/column piece of metadata, right?  I wonder if we should be thinking about creating that abstraction instead (column metadata) instead of baking this specific use of column metadata (provenance) into our data structures and naming.

Yes it is. While it is good to keep future direction in mind i don't want to build internal interfaces that are more general than what we need right now and/or take us in step along the path we know we want to follow.

> 2. On that note, test_positivity already has kinda' 2 forms of provenance, right? a) what data source did the data come from ("X dashboard via covid-tracking"), and b) which calculation did we use to calculate the final metric value ("positive_tests_viral / total_tests_viral").  Do you have a sense of how we would represent that?  If we have a single provenance field per location/column, do we have to come up with some sort of string-encoding of all of that (a JSON blob or something)?

Totally agree that provenance can easily have N sources and 1 method to combine them. JSON blob isn't a bad idea because it is easy to imagine that `source` could be broken down to dataset name and version (for example sha in the github world). That's beyond the scope of this PR.

> Also, just FYI- When I run this branch locally right now I get tons of "Unable to find TestPositivityRatioMethod location_id=iso1:us#iso2:us-sd#fips:46003" errors which I don't get in master. But maybe that'll be fixed by a merge.
Opps, I think this is fixed now.

File: api/can_api_v2_definition.py
Comment (ghop02) [inline api/can_api_v2_definition.py:64]: My first thought is that this would be better served as `testPositivityRatioDetails` with the `method` being the only item inside.  

I think it'll make it more future proof.  If we ever add another metadata option (don't see one immediately, but wouldn't shock me) it will make it easier to make a non breaking change.  In that case, I would prefer to have it in its own object rather than cluttering the top level metrics space.  Additionally it copies an existing pattern of `icuHeadroomDetails`.

File: api/can_api_definition.py
Comment (ghop02) [inline api/can_api_definition.py:83]: Valorum feels pretty unhelpful for anyone who doesn't know the internals of our API data pipeline. I guess there's not really much of a better option right now

File: libs/top_level_metrics.py
Comment (ghop02) [inline libs/top_level_metrics.py:116]: Can you add a comment about when this wouldn't ever be true? after reading into it a bit more I can guess, but some context would be helpful

File: libs/top_level_metrics.py
Comment (ghop02) [inline libs/top_level_metrics.py:160]: Info feels like a dangerous move here.  Do we ever expect this to happen?

File: libs/pipelines/api_v2_pipeline.py
Comment (TomGoBravo) [reply libs/pipelines/api_v2_pipeline.py:123]: That's funny because I saw the implicit passing around of a bound logger as an anti-pattern. What we have right now is not very consistent :-( which is why I've been erring towards keeping everything explicit on the stock. Cleaning up logging to be more consistent seems like a fine idea but for now I'd like to avoid introducing yet another way to do logging in this repo.

File: libs/top_level_metrics.py
Comment (TomGoBravo) [reply libs/top_level_metrics.py:160]: Do you mean you'd like this to be warning so we are less likely to overlook it breaking? SGTM.

File: libs/pipelines/api_v2_pipeline.py
Comment (ghop02) [reply libs/pipelines/api_v2_pipeline.py:123]: Heh, that's why I'm still not sold on structlog :) never really had a problem with the std logger before haha

File: api/can_api_v2_definition.py
Comment (mikelehen) [reply api/can_api_v2_definition.py:137]: FWIW- I'd advocate we don't add it to the API at all until we have specced out the desired API a bit better.

Comment (mikelehen) [reply api/can_api_v2_definition.py:137]: (e.g. as @ghop02 pointed out, "Valorum" isn't what we want to surface in the API)

Comment (ghop02) [reply api/can_api_v2_definition.py:137]: Do we have plans to use it in the short term? I may have been the person to push for this in the API (honestly not 100% sure), but would be good to figure out how to communicate this difference somewhere

Comment (mikelehen) [reply api/can_api_v2_definition.py:137]: Actually...  yes, maybe.  I would like to get CMS data in as a backfill this week, and the CMS data is a 14-day average that only has weekly datapoints, and may be up to 7-10 days lagged...  And we probably need to surface that on the site somehow.

I could do that in a hacky way (e.g. hardcode where we're using CMS data or "detect" it based on the cadence of data points)...  or we could do it propely and plumb it through with this provenance work.

The part that makes me nervous is:
* Exposing "Valorum" as a name to API consumers.
* "Valorum" vs "HHSTesting" isn't sufficient, since we also use Covid Tracking Project for ~3 states right now.
* As Chris mentioned, we may want to nest this under a "testPositivityRatioDetails" object so that we give ourselves flexibility to expand.

So it'd be nice if we could expose this as a "private" API until we are happy with it so we don't have to worry about breaking people..  though, in practice probably nobody will notice / use this data, so it may be okay for us to make breaking changes.

File: api/can_api_v2_definition.py
Comment (mikelehen) [reply api/can_api_v2_definition.py:64]: testPositivityRatioDetails =
```
{
  source: 'CMS'
}
```

File: libs/top_level_metrics.py
Comment (TomGoBravo) [reply libs/top_level_metrics.py:116]: I probably made this function work when provenance is None before I had loading the provenance done correctly, then left it in place because it seemed better to log a warning than raise an exception. Would you prefer an exception?

File: libs/top_level_metrics.py
Comment (ghop02) [inline libs/top_level_metrics.py:120]: After looking at the logs, looks like this is pretty noisy.  We should probably remove this log or maybe find a way to log total number of "other" results

File: libs/top_level_metrics.py
Comment (ghop02) [inline libs/top_level_metrics.py:265]: 🙏

**Review by ghop02 — APPROVED** [review]
would like to clean the logging up but looks good!

File: libs/top_level_metrics.py
Comment (TomGoBravo) [reply libs/top_level_metrics.py:120]: Cleaning this up will be much easier when test positivity calculation is extracted from top_level_metrics. Changed to debug for now.

**Pull request merged**
